include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -0,0 +1,2779 @@
+      return mlir::failure();
+    result.addAttribute(fieldAttrName(), builder.getStringAttr(fieldName));
+    if (!recty.dyn_cast<RecordType>())
+      return mlir::failure();
+    result.addAttribute(typeAttrName(), mlir::TypeAttr::get(recty));
+    if (!parser.parseOptionalLParen()) {
+      llvm::SmallVector<mlir::OpAsmParser::OperandType, 8> operands;
+      llvm::SmallVector<mlir::Type, 8> types;
+      auto loc = parser.getNameLoc();
+      if (parser.parseOperandList(operands,
+                                  mlir::OpAsmParser::Delimiter::None) ||
+          parser.parseRParen() ||
+          parser.parseColonTypeList(types) ||
+          parser.resolveOperands(operands, types, loc, result.operands))
+        return mlir::failure();
+    }
+    mlir::Type fieldType = fir::FieldType::get(builder.getContext());
+    if (parser.addTypeToList(fieldType, result.types))
+      return mlir::failure();
+    return mlir::success();
+  }];
+
+  let printer = [{
+    p << getOperationName() << ' '
+      << getAttrOfType<mlir::StringAttr>(fieldAttrName()).getValue() << ", "
+      << getAttr(typeAttrName());
+    if (getNumOperands()) {
+      p << '(';
+      p.printOperands(lenparams());
+      auto sep = ") : ";
+      for (auto op : lenparams()) {
+        p << sep;
+        if (op)
+          p.printType(op.getType());
+        else
+          p << "()";
+        sep = ", ";
+      }
+    }
+  }];
+
+  let builders = [OpBuilder<
+    "Builder *builder, OperationState &result, StringRef fieldName,"
+    "Type recTy, ArrayRef<Value> operands = {}",
+    [{
+      result.addAttribute(fieldAttrName(), builder->getStringAttr(fieldName));
+      result.addAttribute(typeAttrName(), TypeAttr::get(recTy));
+      result.addOperands(operands);
+    }]
+  >];
+
+  let extraClassDeclaration = [{
+    static constexpr llvm::StringRef fieldAttrName() { return "field_id"; }
+    static constexpr llvm::StringRef typeAttrName() { return "on_type"; }
+  }];
+}
+
+def fir_GenDimsOp : fir_OneResultOp<"gendims", [NoSideEffect]> {
+
+  let summary = "generate a value of type `!fir.dims`";
+
+  let description = [{
+    The arguments are an ordered list of integral type values that is a
+    multiple of 3 in length.  Each such triple is defined as: the lower
+    index, the extent, and the stride for that dimension. The dimension
+    information is given in the same row-to-column order as Fortran. This
+    abstract dimension value must describe a reified object, so all dimension
+    information must be specified.  The extent must be nonnegative and the
+    stride must not be zero.
+
+      %d = fir.gendims %l, %u, %s : (index, index, index) -> !fir.dims<1>
+  }];
+
+  let arguments = (ins Variadic<AnyIntegerType>:$triples);
+
+  let results = (outs fir_DimsType);
+
+  let assemblyFormat = [{
+    operands attr-dict `:` functional-type(operands, results)
+  }];
+
+  let verifier = [{
+    auto size = triples().size();
+    if (size < 1 && size <= 16 * 3)
```

Comment:
I don't quite parse this, it seems like `rhs` would always be true if `lhs` is true.

Reply:
Sleep deprivation. Fixed.